docs/debug.md

```diff
@@ -0,0 +1,48 @@
+# CMake Debugging
+
+Starting with CMake 3.27, debugging CMake is now supported in CMake Tools!
+
+The following documentation will help you understand the various ways you can debug CMake.
+
+## Debugging from CMake Tools UI entry points
+
+The most common reason to debug CMake is to debug CMake Configuration! There are many ways that you can accomplish this:
+
+* Commands
+  * CMake: Configure with CMake Debugger
+  * CMake: Delete Cache and Reconfigure with CMake Debugger
+* Folder Explorer
+  * Right click on CMakeLists.txt context menu entry.
+* Project Outline
+  * Right click on CMakeLists.txt context menu entry.
+  * Expand the "..." in the project outline. There is an entry to use the Debugger!
+
+## Debugging from CMake Tools launch.json
+
+CMake Tools now provides a new debug type `cmake`!
+
+The `cmake` debug type supports three different types of `cmakeDebugType`: `configure`, `external`, `script`. They each come with their own settings that can be used to modify and control the debug session!
+
+Listed below are the settings that are available for each configuration based on `cmakeDebugType`:
+
+* `configure`
+  * required
+  * optional
+    * `pipeName` - Name of the pipe (on Windows) or domain socket (on Unix) to use for debugger communication.
+    * `clean` - Clean prior to configuring.
+    * `configureAll` - Configure for all projects.
+    * `dapLog` - Where the debugger DAP log should be logged.
+* `external`
+  * required
+    * `pipeName` - Name of the pipe (on Windows) or domain socket (on Unix) to use for debugger communication.
+  * optional
```

Comment:
can this get an optional ``command` : []` field/array? So that VS code executes the command I would normally use to start the cmake pipe externally?

Reply:
To help me think through the possible design implications, could you go into more detail regarding your use case for the `external` mode? 

What kind of external processes are you expecting? It is a CMake invocation? Or is it invoking something else that internally invokes CMake?

Thanks!

Reply:
> It is a CMake invocation? Or is it invoking something else that internally invokes CMake?

Could be both of those. 

> What are you expecting or hoping to use in the "command" field?

For example I could write a (CMake) script (not the debug target) which does the following:
- FetchContent vcpkg
- bootstrap vcpkg
- vcpkg install (with some additional options to pass on the debugger pipe info)
- vcpkg internally calls `cmake -P` which could be debugged
- vcpkg internally calls `cmake` in configure mode which could be debugged

The intention of having a `command` field is to automate the process. E.g. 
`"command" : ["cmake", "-P", "my-internal-cmake-caller-script.cmake", "-DDEBUGGER_PIPE_NAME=<name_of_debugger_pipe>"] `
`DEBUGGER_PIPE_NAME` is internally passed on to some process calling cmake. (doesn't really matter what it is)

But if "command" is added "workingDir" probably also makes sense to define where the command is executed.

Reply:
That being said I am also happy with the current approach. It is just that it takes more manual labor for startup.

Reply:
Understood. I wonder if using a `preLaunchTask` is what you want. This way the pre launch task can be fully arbitrary. Do you think using that would suit your needs?

Reply:
Don't know about that. If there is a possibility to detach then probably yes if it needs to return than probably no. Need to read up on that first.

Reply:
I just tested and found that if you set up a tasks.json task with the following settings as part of that task: 

```
"isBackground": true,
        // This task is run before some debug tasks.
        // Problem is, it's a watch script, and since it never exits, VSCode
        // complains. All this is needed so VSCode just lets it run.
        "problemMatcher": [
            {
            "pattern": [
                {
                "regexp": ".",
                "file": 1,
                "location": 2,
                "message": 3
                }
            ],
            "background": {
                "activeOnStart": true,
                "beginsPattern": ".",
                "endsPattern": ".",
            }
            }
        ]
```
then you can successfully have a preLaunch task.

Reply:
Hmm I need some delay or some retry count to connect the debugger to the pipe since the debugger wants to start before the prelaunch task reached the point where the pipe is opened.

Reply:
OK I think there is nothing to do here. I can just define a task and let it run via VS Code directly instead of manually typing it. Never used task since I mainly use CMake presets. Would be nice if there is a way to somehow get the pipename from a task.json via some variable.